src/test/java/com/emc/object/s3/S3JerseyClientTest.java

```diff
@@ -2971,4 +2972,172 @@ protected void assertAclEquals(AccessControlList acl1, AccessControlList acl2) {
+
+        // Only the particular version of the object should get deleted and no other versions of object should be affected
+        client.deleteObject(new DeleteObjectRequest(bucketName, key).withVersionId(versionId2));
+        Assert.assertEquals(2, client.getObject(new GetObjectRequest(bucketName, key).withVersionId(versionId1), String.class).getObjectMetadata().getTaggingCount());
+
+        // Object and associated multiple tags should get deleted
+        Assert.assertEquals(2, client.getObject(new GetObjectRequest(bucketName, key).withVersionId(versionId1), String.class).getObjectMetadata().getTaggingCount());
+        client.deleteObject(new DeleteObjectRequest(bucketName, key).withVersionId(versionId1));
+        try {
+            client.getObject(new GetObjectRequest(bucketName, key).withVersionId(versionId1), String.class);
+        } catch (S3Exception e) {
+            Assert.assertEquals(404, e.getHttpCode());
+            Assert.assertEquals("NoSuchKey", e.getErrorCode());
+        }
+
+    }
+
+    @Test
+    public void testCopyObjectWithTagging() {
+
+        // set up env
+        String bucketName = getTestBucket(), key1 = "test-object-tagging-src", key2 = "test-object-tagging-dest1", key3 = "test-object-tagging-dest2", content = "Hello Object Tagging!";
+
+        // should be able to copy the object and copied object should have the tags also
+        client.putObject(new PutObjectRequest(bucketName, key1, content)
+                .withObjectTagging(new ObjectTagging().withTagSet(Arrays.asList(new ObjectTag("k11", "v11"), new ObjectTag("k12", "v12")))));
+        client.copyObject(new CopyObjectRequest(bucketName, key1, bucketName, key2));
+        client.copyObject(new CopyObjectRequest(bucketName, key1, bucketName, key2));
+        Assert.assertEquals(2, client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key2)).getTagSet().size());
+
+        // Versioned object should be copied and user should be able to get the same along with tags
+        client.setBucketVersioning(bucketName, new VersioningConfiguration().withStatus(VersioningConfiguration.Status.Enabled));
+        client.putObject(new PutObjectRequest(bucketName, key1, content)
+                .withObjectTagging(new ObjectTagging().withTagSet(Arrays.asList(new ObjectTag("k11", "v11"), new ObjectTag("k12", "v12")))));
+        String versionId = client.listVersions(bucketName, null).getVersions().get(0).getVersionId();
+        client.copyObject(new CopyObjectRequest(bucketName, key1, bucketName, key2).withSourceVersionId(versionId));
+        Assert.assertEquals(2, client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key2).withVersionId(versionId)).getTagSet().size());
+
+    }
+
+    @Test
+    public void testMultipartUploadWithTagging() {
+
+        // set up env
+        String bucketName = getTestBucket(), key = "test-object-tagging";
+        int fiveMB = 5 * 1024;
+        byte[] content = new byte[11 * 1024];
```

Comment:
I understand there are already test cases written in a similar way, however, it might be better to define part size and total size for flexibility. Now the contents of is1 and is2 are the same, but it will be different and clearer to do data random, upload parts, consolidate parts together in loop.